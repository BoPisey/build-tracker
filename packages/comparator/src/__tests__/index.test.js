import BuildComparator, * as Comparator from '../';

const build1 = {
  meta: { revision: '1234567' },
  artifacts: {
    churros: { hash: 'abc', size: 456, gzipSize: 90 },
    tacos: { hash: 'abc', size: 123, gzipSize: 45 }
  }
};

const build2 = {
  meta: { revision: '8901234' },
  artifacts: {
    burritos: { hash: 'def', size: 469, gzipSize: 93 },
    tacos: { hash: 'abc', size: 123, gzipSize: 45 }
  }
};

const build3 = {
  meta: { revision: '5678901' },
  artifacts: {
    burritos: { hash: 'ghi', size: 345, gzipSize: 85 },
    tacos: { hash: 'abc', size: 123, gzipSize: 45 }
  }
};

describe('BuildComparator', () => {
  describe('artifactNames', () => {
    test('includes all artifact names', () => {
      const comparator = new BuildComparator([build1, build2]);
      expect(comparator.artifactNames).toEqual(expect.arrayContaining(['burritos', 'churros', 'tacos']));
    });
  });

  describe('buildDeltas', () => {
    test('returns an array of deltas', () => {
      const comparator = new BuildComparator([build1, build2]);
      expect(comparator.buildDeltas).toMatchSnapshot();
    });
  });

  describe('matrix', () => {
    test('includes a header', () => {
      const comparator = new BuildComparator([build1, build2]);
      expect(comparator.matrix.header).toEqual(comparator.matrixHeader);
      expect(comparator.matrix.header).toMatchSnapshot();
    });

    test('includes a total', () => {
      const comparator = new BuildComparator([build1, build2]);
      expect(comparator.matrix.total).toEqual(comparator.matrixTotal);
      expect(comparator.matrix.total).toMatchSnapshot();
    });

    test('includes a body of all artifacts', () => {
      const comparator = new BuildComparator([build1, build2]);
      expect(comparator.matrix.body).toEqual(comparator.matrixBody);
      expect(comparator.matrix.body).toMatchSnapshot();
    });
  });

  describe('getAscii', () => {
    test('gets an ASCII table', () => {
      const comparator = new BuildComparator([build1, build2]);
      expect(comparator.getAscii()).toMatchSnapshot();
    });

    test('can filter rows', () => {
      const comparator = new BuildComparator([build1, build2]);
      const rowFilter = row => {
        return row.some(cell => {
          if ('sizePercent' in cell && 'size' in cell) {
            return cell.size > 3;
          }
          return false;
        });
      };
      expect(comparator.getAscii({ rowFilter })).toMatchSnapshot();
    });
  });

  describe('getCsv', () => {
    test('gets a CSV formatted table', () => {
      const comparator = new BuildComparator([build1, build2]);
      expect(comparator.getCsv()).toMatchSnapshot();
    });

    test('can filter rows', () => {
      const comparator = new BuildComparator([build1, build2]);
      const rowFilter = row => {
        return row.some(cell => {
          if ('sizePercent' in cell && 'size' in cell) {
            return cell.size > 3;
          }
          return false;
        });
      };
      expect(comparator.getCsv({ rowFilter })).toMatchSnapshot();
    });
  });
});
