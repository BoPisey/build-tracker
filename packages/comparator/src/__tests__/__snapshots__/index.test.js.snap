// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildComparator buildDeltas returns an array of deltas 1`] = `
Array [
  Object {
    "artifactDeltas": Array [],
    "deltas": Array [],
    "meta": Object {
      "revision": "1234567",
    },
  },
  Object {
    "artifactDeltas": Array [
      Object {
        "burritos": Object {
          "gzipSize": 93,
          "gzipSizePercent": 1,
          "hashChanged": true,
          "size": 469,
          "sizePercent": 1,
          "type": "delta",
        },
        "churros": Object {
          "gzipSize": -90,
          "gzipSizePercent": 1,
          "hashChanged": true,
          "size": -456,
          "sizePercent": 1,
          "type": "delta",
        },
        "tacos": Object {
          "gzipSize": 0,
          "gzipSizePercent": 1,
          "hashChanged": false,
          "size": 0,
          "sizePercent": 1,
          "type": "delta",
        },
      },
    ],
    "deltas": Array [
      Object {
        "gzipSize": 3,
        "gzipSizePercent": 1.0222222222222221,
        "size": 13,
        "sizePercent": 1.0224525043177892,
        "type": "totalDelta",
      },
    ],
    "meta": Object {
      "revision": "8901234",
    },
  },
]
`;

exports[`BuildComparator getAscii can filter rows 1`] = `
".----------------------------------------.
|          | 1234567 | 8901234 |   Δ0    |
|----------|---------|---------|---------|
| burritos | 0       | 93      | 93 (1%) |
'----------------------------------------'"
`;

exports[`BuildComparator getAscii gets an ASCII table 1`] = `
".-----------------------------------------.
|          | 1234567 | 8901234 |    Δ0    |
|----------|---------|---------|----------|
| burritos | 0       | 93      | 93 (1%)  |
| churros  | 90      | 0       | -90 (1%) |
| tacos    | 45      | 45      | 0 (1%)   |
'-----------------------------------------'"
`;

exports[`BuildComparator getCsv can filter rows 1`] = `
",1234567,8901234,Δ0
burritos,0,93,93 (1%)"
`;

exports[`BuildComparator getCsv gets a CSV formatted table 1`] = `
",1234567,8901234,Δ0
burritos,0,93,93 (1%)
churros,90,0,-90 (1%)
tacos,45,45,0 (1%)"
`;

exports[`BuildComparator matrix includes a body of all artifacts 1`] = `
Array [
  Array [
    Object {
      "text": "burritos",
      "type": "artifact",
    },
    Object {
      "gzipSize": 0,
      "size": 0,
      "type": "total",
    },
    Object {
      "gzipSize": 93,
      "size": 469,
      "type": "total",
    },
    Object {
      "gzipSize": 93,
      "gzipSizePercent": 1,
      "hashChanged": true,
      "size": 469,
      "sizePercent": 1,
      "type": "delta",
    },
  ],
  Array [
    Object {
      "text": "churros",
      "type": "artifact",
    },
    Object {
      "gzipSize": 90,
      "size": 456,
      "type": "total",
    },
    Object {
      "gzipSize": 0,
      "size": 0,
      "type": "total",
    },
    Object {
      "gzipSize": -90,
      "gzipSizePercent": 1,
      "hashChanged": true,
      "size": -456,
      "sizePercent": 1,
      "type": "delta",
    },
  ],
  Array [
    Object {
      "text": "tacos",
      "type": "artifact",
    },
    Object {
      "gzipSize": 45,
      "size": 123,
      "type": "total",
    },
    Object {
      "gzipSize": 45,
      "size": 123,
      "type": "total",
    },
    Object {
      "gzipSize": 0,
      "gzipSizePercent": 1,
      "hashChanged": false,
      "size": 0,
      "sizePercent": 1,
      "type": "delta",
    },
  ],
]
`;

exports[`BuildComparator matrix includes a header 1`] = `
Array [
  Object {
    "text": "",
    "type": "text",
  },
  Object {
    "revision": "1234567",
    "type": "revision",
  },
  Object {
    "revision": "8901234",
    "type": "revision",
  },
  Object {
    "againstRevision": "1234567",
    "deltaIndex": 0,
    "revision": "8901234",
    "type": "revisionDelta",
  },
]
`;

exports[`BuildComparator matrix includes a total 1`] = `
Array [
  Object {
    "text": "All",
    "type": "artifact",
  },
  Object {
    "gzipSize": 135,
    "size": 579,
    "type": "total",
  },
  Object {
    "gzipSize": 138,
    "size": 592,
    "type": "total",
  },
  Object {
    "gzipSize": 3,
    "gzipSizePercent": 1.0222222222222221,
    "size": 13,
    "sizePercent": 1.0224525043177892,
    "type": "totalDelta",
  },
]
`;
